extensions/amp-video/0.1/amp-video.js

```diff
@@ -83,18 +90,47 @@ class AmpVideo extends AMP.BaseElement {
    * @override
    */
   preconnectCallback(opt_onLayout) {
-    const videoSrc = this.getVideoSource_();
+    const videoSrc = this.getVideoSourceForPreconnect_();
     if (videoSrc) {
       assertHttpsUrl(videoSrc, this.element);
       this.preconnect.url(videoSrc, opt_onLayout);
     }
   }
 
+  /**
+   * @override
+   *
+   * @overview
+   * AMP Cache may selectively cache certain video sources (based on size, type
+   * etc..). When AMP Cache does so, it rewrites the `src` for `amp-video` and
+   * `source` children that are cached and adds a `amp-orig-src` attribute
+   * pointing to the original source.
+   *
+   * There are two separate runtime concerns that we handle here:
+   *
+   * 1) Handling 404s
+   * Eventhough AMP Cache rewrites the `src` to point to the CDN, the actual
+   * video may not be ready in the cache yet, in those cases the CDN will
+   * return a 404. Runtime handles this situation by appending an additional
+   * <source> pointing to the original src AFTER the cached source so browser
+   * will automatically proceed to the next source if one fails.
+   *
+   * 2) Prerendering
+   * Now that some sources might be cached, we can preload them during prerender
+   * phase. Runtime handles this by adding any cached sources to the <video>
+   * element during prerender and automatically sets the `preload` to `auto`
+   * so browsers (based on their own heuristics) can start fetching the cached
+   * videos. If `preload` is specified by the author, then it takes precedence.
```

Comment:
Something maybe that we can revisit later, probably low impact anyway, but it feels like that in prerendering mode we could feel free to impose preload independently of the preexisting value?

E.g., for images, we always prerender them no matter what. At the same time, I imagine this is rarely used.

Reply:
Images have an special `noprerender` attribute that can be used to optout. We would need a similar optout for videos which for now is `preload=none`.